File: 10-11-2025-Stock_Buy_and_Sell_with_Cooldown.cpp
class Solution {
  public:
    int maxProfit(vector<int>& arr) {
        // code here
        int n = arr.size();
        
        // base cases
        int buy = -arr[0];  // maximum profit if we buy a stock today
        int sell = 0;   // maximum profit if we sell our holding today
        int skip = 0;   // maximum profit if we skip trade today
        
        for (int i=1; i<n; i++) {
            int price = arr[i];
            
            int prvBuy = buy;
            int prvSell = sell;
            int prvSkip = skip;
            
            buy = max(prvBuy, prvSkip-price);  // maximum of previous buying or today's buying
            sell = prvBuy+price;    // profit after selling today
            skip = max(prvSkip, prvSell);   // max of previous skipping or skipping after yesterday sell
        }
        
        return max(sell, skip);
    }
};

File: 10-11-2025-Stock_Buy_and_Sell_with_Cooldown.java
class Solution {
    public int maxProfit(int[] arr) {
        int n = arr.length;

        int buy = -arr[0];  // max profit if we buy a stock today
        int sell = 0;       // max profit if we sell today
        int skip = 0;       // max profit if we skip today

        for (int i = 1; i < n; i++) {
            int price = arr[i];

            int prvBuy = buy;
            int prvSell = sell;
            int prvSkip = skip;

            buy = Math.max(prvBuy, prvSkip - price);
            sell = prvBuy + price;
            skip = Math.max(prvSkip, prvSell);
        }

        return Math.max(sell, skip);
    }
}
